src/main/java/seedu/address/logic/CommandHistory.java

```diff
@@ -0,0 +1,74 @@
+package seedu.address.logic;
+
+import static java.util.Objects.requireNonNull;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+
+/**
+ * Stores the history of commands executed. (Solution inspired from addressbook-level-4)
+ */
+public class CommandHistory {
+    private final ObservableList<String> userInputHistory = FXCollections.observableArrayList();
+    private final ObservableList<String> unmodifiableUserInputHistory =
+            FXCollections.unmodifiableObservableList(userInputHistory);
+
+    private int currentIndex = 0;
+
+    /**
+     * Appends {@code userInput} to the list of user input entered.
+     */
+    public void add(String userInput) {
+        requireNonNull(userInput);
+        userInputHistory.add(userInput);
+        currentIndex = userInputHistory.size(); // reset index to the end
+    }
+
+    /**
+     * Returns an unmodifiable view of {@code userInputHistory}.
+     */
+    public ObservableList<String> getHistory() {
+        return unmodifiableUserInputHistory;
+    }
+
+    /**
+     * Returns the previous command in the history.
+     */
+    public String getPrevious() {
+        if (currentIndex > 0) {
+            currentIndex--;
+        }
+        return userInputHistory.get(currentIndex);
+    }
+
+    /**
+     * Returns the next command in the history.
+     */
+    public String getNext() {
+        if (currentIndex <= userInputHistory.size() - 1) {
```

Comment:
Can consider not having it show blank line when reach the latest command, or to have the result display message saying no latest command.

Reply:
Thanks for the suggestion! I'll use the display to feedback to the user.